core/src/main/java/org/commonjava/maven/galley/io/SpecialPathManagerImpl.java

```diff
@@ -15,61 +15,129 @@
+import static org.commonjava.maven.galley.io.SpecialPathConstants.MVN_SP_PATH_SET;
+
 /**
  * Created by jdcasey on 1/27/16.
  */
 @ApplicationScoped
 public class SpecialPathManagerImpl
         implements SpecialPathManager
 {
-    private List<SpecialPathInfo> specialPaths;
+    private static final Logger logger = LoggerFactory.getLogger( SpecialPathManagerImpl.class );
+
+    private List<SpecialPathInfo> stdSpecialPaths;
+
+    private Map<String, SpecialPathSet> pkgtypes;
 
     public SpecialPathManagerImpl()
     {
-        specialPaths = new ArrayList<>();
-        specialPaths.addAll( SpecialPathConstants.STANDARD_SPECIAL_PATHS );
+        initPkgPathSets();
+    }
+
+    @PostConstruct
+    public void initPkgPathSets()
+    {
+        stdSpecialPaths = new ArrayList<>();
+        stdSpecialPaths.addAll( SpecialPathConstants.STANDARD_SPECIAL_PATHS );
+        pkgtypes = new ConcurrentHashMap<>(  );
+        pkgtypes.put( MVN_SP_PATH_SET.getPackageType(), MVN_SP_PATH_SET );
     }
 
     @Override
-    public synchronized void registerSpecialPathInfo( SpecialPathInfo pathInfo )
+    public void registerSpecialPathInfo( SpecialPathInfo pathInfo )
     {
-        specialPaths.add( pathInfo );
+        //FIXME: do we need a flag of "isStandard" to add into stdSpecialPaths?
+        registerSpecialPathInfo( pathInfo, PKG_TYPE_MAVEN );
+    }
+
+    @Override
+    public void registerSpecialPathInfo( SpecialPathInfo pathInfo, final String pkgType )
+    {
+        for ( Map.Entry<String, SpecialPathSet> entry : pkgtypes.entrySet() )
+        {
+            if ( pkgType.equals( entry.getKey() ) )
+            {
+                entry.getValue().registerSpecialPathInfo( pathInfo );
+            }
+        }
     }
 
     @Override
     public synchronized void deregisterSpecialPathInfo( SpecialPathInfo pathInfo )
     {
-        specialPaths.remove( pathInfo );
+        //FIXME: do we need a flag of "isStandard" to remove from stdSpecialPaths?
+        deregisterSpecialPathInfo( pathInfo, PKG_TYPE_MAVEN );
+    }
+
+    @Override
+    public void deregisterSpecialPathInfo( SpecialPathInfo pathInfo, String pkgType )
+    {
+        for ( Map.Entry<String, SpecialPathSet> entry : pkgtypes.entrySet() )
+        {
+            if ( pkgType.equals( entry.getKey() ) )
+            {
+                entry.getValue().deregisterSpecialPathInfo( pathInfo );
+            }
+        }
+    }
+
+    @Override
+    public void registerSpecialPathSet( SpecialPathSet pathSet )
+    {
+        if ( pkgtypes.containsKey( pathSet.getPackageType() ) )
+        {
+            logger.warn(
+                    "[Galley] The package types already contains the path set for this package type {}, will override it",
+                    pathSet.getPackageType() );
+        }
+
+        pkgtypes.put( pathSet.getPackageType(), pathSet );
+    }
+
+    @Override
+    public void deregesterSpecialPathSet( SpecialPathSet pathSet )
```

Comment:
IMO we should return the removed SpecialPathSet. It's similar to Map semantics, and if people don't need the return value they can just discard it.